storage/innobase/xtrabackup/test/t/xbcloud.sh

```diff
@@ -165,8 +169,111 @@ if  grep -q "Upload failed" $topdir/pxb-2279.log ; then
     die 'xbcloud exit with error on upload'
 fi
 
+# PXB-2112 - Support for storage class
+test_storage_class=false
+
+if grep -q "storage=s3" $topdir/xbcloud.cnf ; then
+  vlog "Testing S3 storage class"
+  test_storage_class=true
+  storage_class_parameter="--s3-storage-class=GLACIER"
+elif grep -q "storage=google" $topdir/xbcloud.cnf ; then
+  vlog "Testing GC storage class"
+  test_storage_class=true
+  storage_class_parameter="--google-storage-class=COLDLINE"
+fi
+
+if grep -q "minioadmin" $topdir/xbcloud.cnf ; then
+  test_storage_class=false
+fi
+
+if [ "$test_storage_class" = true ]; then
+  run_cmd touch $topdir/xtrabackup_info
+  cd $topdir/
+  xbstream -c xtrabackup_info | \
+  xbcloud --defaults-file=$topdir/xbcloud.cnf ${storage_class_parameter} \
+   put ${storage_class_folder} 2> $topdir/storage_class.log
+
+ if  grep -q "expected objects using storage class" $topdir/storage_class.log ; then
+    xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+     ${storage_class_folder}
+     die 'xbcloud did not upload using correct storage class'
+ fi
+ if  grep -q "Upload failed" $topdir/storage_class.log ; then
+   xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+     ${storage_class_folder}
+     die 'xbcloud exit with error on upload'
+ fi
+ run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+  ${storage_class_folder}
+fi
+
 # cleanup
 run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
 	${full_backup_name} --parallel=4
 run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
 	${inc_backup_name} --parallel=4
+rm -rf ${full_backup_dir}
+rm -rf $topdir/downloaded_full
+
+
+#check if password is stored in backup directory or in ps output
+
+## grep for all key in secret_key
+params=($(echo $XBCLOUD_CREDENTIALS | tr ";" "\n"))
+secret_keys=();
+for param in "${params[@]}"
+do
+ if [[ $param =~ "key=" ]]; then
+	 secret_keys+=(`echo $param | awk -F 'key=' '{print $2}'`)
+ fi
+done
+
+### create table in MyISAM to let backup wait until there is pending update
+mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
+mysql -e "INSERT into t1 values(100)" test
+
+### The running update on myisam table will hold access to the table.
+### PXB will do FLUSH TABLES on MyISAM tables which requires access
+### to the table and waits for the running update on MyISAM to finish.
+mysql -e "SELECT SLEEP(1000) from t1 for update " test &
+
+xtrabackup --backup --stream=xbstream --extra-lsndir=$full_backup_dir \
+	   --target-dir=$full_backup_dir | \
+    run_cmd xbcloud $XBCLOUD_CREDENTIALS put \
+	    --parallel=4 \
+	    ${full_backup_name} &
+job_pid=$!
+
+#sleep is to ensure password is removed by xbcloud
+sleep 10
```

Comment:
Is this still necessary? can we perhaps grep the log for an indication that the backup has started/passed the argument handling part?

Reply:
Which logs to grep? with backup format being xbstream it won't logged anything on disk.

Reply:
log messages are sent to stderr, redirects ( | or >) will redirect only STDOUT. We are still writing stderr to `$OUTFILE`

Reply:
with PS. Update on MYISAM blocks LTFB.  not sure what to check if xbcloud has already parsed the value or not. 

error: http request failed: Couldn't resolve host name
error: http request failed: Couldn't resolve host name
220406 14:47:49 xbcloud: Successfully connected.
2022-04-06T14:47:49.315246+05:30 0 [Note] [MY-011825] [Xtrabackup] recognized client arguments: --target-dir=/Users/rahulmalik/MySQL/data/xp8 --core-file --user=root --socket=/tmp/socket_3208.sock --keyring-file-data=/Users/rahulmalik/MySQL/data/p8/key.key --xtrabackup-plugin-dir=/Users/rahulmalik/MySQL/src/x8/bld/plugin_output_directory --backup=1 --stream=xbstream
/Users/rahulmalik/MySQL/src/x8/bld/runtime_output_directory/xtrabackup version 8.0.28-20 based on MySQL server 8.0.28 macos12.2 (x86_64) (revision id: b980f66bc03)
220406 14:47:49  version_check Connecting to MySQL server with DSN 'dbi:mysql:;mysql_read_default_group=xtrabackup;mysql_socket=/tmp/socket_3208.sock' as 'root'  (using password: NO).
Failed to connect to MySQL server as DBD::mysql module is not installed at - line 1548.
2022-04-06T14:47:49.442151+05:30 0 [Note] [MY-011825] [Xtrabackup] Connecting to MySQL server host: localhost, user: root, password: not set, port: not set, socket: /tmp/socket_3208.sock
2022-04-06T14:47:49.454317+05:30 0 [Note] [MY-011825] [Xtrabackup] Using server version 8.0.28-19-debug
2022-04-06T14:47:49.488822+05:30 0 [Note] [MY-011825] [Xtrabackup] Executing LOCK TABLES FOR BACKUP ...

Reply:
> 220406 14:47:49 xbcloud: Successfully connected.

This is an indication that we have parsed everything and managed to connect to backend storage and are ready to upload. At this point no sensitive data should be exposed at `ps -ef`